Migrate Enzyme tests to React Testing Library

jeffstieler opened the pull request:
This PR migrates all remaining Enzyme tests to React Testing Library.

Where reasonable, it also changes the test behavior to check user facing functionality rather than component props or specific HTML elements.

Heads up to @c-shultz @jconroy @allendav - this PR changes test code from your teams. Please have someone review!

### Detailed test instructions:

- Verify the JS tests pass
- Review the new test code

<!--- Note: When displaying information based on sample data, such as SwaggerHub, 
be sure to detail parts affected in Release Notes --->

### Changelog Note:

<!--- Optional: Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance:. If no note is entered, one will be constructed from the title and labels. --->

Dev: Remove Enzyme in favor of React Testing Library

jeffstieler committed:
Migrate leaderboard tests to RTL.

jeffstieler committed:
Remove test of default prop value.

jeffstieler committed:
Migrate ReportSummary tests to RTL.

jeffstieler committed:
Migrate ActivityCard tests to RTL.

jeffstieler committed:
Migrate ActivityCardPlaceholder tests to RTL.

jeffstieler committed:
Migrate remaining ProductType tests to RTL.

jeffstieler committed:
Migrate Card tests to RTL.

jeffstieler committed:
Update RTL and user event packages.

jeffstieler committed:
Migrate Date tests to RTL.

jeffstieler committed:
Migrate D3Legend tests to RTL.

jeffstieler committed:
Migrate D3Base tests to RTL.

jeffstieler committed:
Migrate Gravatar tests to RTL.

jeffstieler committed:
Migrate ImageUpload tests to RTL.

jeffstieler committed:
Migrate ProductImage tests to RTL.

jeffstieler committed:
Migrate Rating tests to RTL.

jeffstieler committed:
Migrate Search tests to RTL.

jeffstieler committed:
Migrate Plugins tests to RTL.

jeffstieler committed:
Migrate SelectControl tests to RTL.

jeffstieler committed:
Migrate Timeline tests to RTL.

Remove tests that inspect DOM since there are snapshots.

jeffstieler committed:
Migrate DismissModal tests to RTL.

jeffstieler committed:
Migrate SetupNotice tests to RTL.

jeffstieler committed:
Migrate WelcomeCard tests to RTL.

jeffstieler committed:
Fix setup error reason retrieval in ShippingBanner.

jeffstieler committed:
Migrate ShippingBanner tests to RTL.

jeffstieler committed:
Migrate RecommendedExtensions tests to RTL.

jeffstieler committed:
Migrate KnowledgeBase tests to RTL.

jeffstieler committed:
Rename enzyme setup file, modify to setup RTL.

jeffstieler committed:
No need to import jest-dom in test files.

jeffstieler committed:
Remove enzyme dependency.

vbelolapotkov commented on packages/components/src/select-control/test/index.js:
```diff
@@ -19,43 +18,43 @@ describe( 'SelectControl', () => {
-					.length
-			).toBe( 3 );
-		}, 0 );
+		expect(
+			getByRole( 'option', { name: 'lorem 1' } )
+		).not.toBeEmptyDOMElement();
+		expect(
+			getByRole( 'option', { name: 'lorem 2' } )
+		).not.toBeEmptyDOMElement();
+		expect(
+			getByRole( 'option', { name: 'bar' } )
+		).not.toBeEmptyDOMElement();
```
Not critical, but iterating over options makes sense to me here: `options.forEach( option => expect( getByRole( 'option', { name: option.label } ) ).toBeInTheDocument() );`

vbelolapotkov reviewed (commented):
Well done @jeffstieler ! It looks great and I haven't noticed anything critical 🎉  

Suggested alternative patterns in some places. They are not mandatory but IMO lead to better readability and tests output in case of failure. Also note, I've suggested most of alternative patterns in a single place but noticed they might be applied in many other places.

Hope that helps 🙂

jeffstieler commented on client/analytics/components/leaderboard/test/index.js:
```diff
@@ -81,28 +83,39 @@ describe( 'Leaderboard', () => {
-		expect( firstRow[ 2 ].value ).toBe( mockData[ 0 ].orders_count );
-		expect( firstRow[ 3 ].value ).toBe(
-			formatDecimal( mockData[ 0 ].net_revenue )
+		const leaderboard = container.querySelector(
```
In my first pass through these tests I was hesitant to change their behavior too much. That said, introducing snapshot testing where we are just checking markup seems perfectly reasonable. Thanks for the nudge!

jeffstieler commented on client/analytics/components/report-chart/test/index.js:
```diff
@@ -1,45 +1,9 @@
-};
-
 describe( 'ReportChart', () => {
-	test( 'should set the time-comparison mode prop by default', () => {
-		const reportChart = mount(
-			<ReportChart
-				path={ path }
-				primaryData={ data }
-				query={ {} }
-				secondaryData={ data }
-				selectedChart={ selectedChart }
-			/>
-		);
-		const chart = reportChart.find( 'Chart' );
-
-		expect( chart.props().mode ).toEqual( 'time-comparison' );
-	} );
```
I believe that DOM changes would result from this test, but they would be in the nested `<Chart>` component, not the `<ReportChart>` that is under test here.

I removed this test case because it felt really low value. It only covers the default case here: https://github.com/woocommerce/woocommerce-admin/blob/7232d514329e54e61711fb3d09d8cbfceb9f356d/client/analytics/components/report-chart/index.js#L355-L358

jeffstieler commented on client/analytics/components/report-summary/test/index.js:
```diff
@@ -49,86 +50,94 @@ describe( 'ReportSummary', () => {
-		expect( summaryNumber.props().prevValue ).toBe( '500.25' );
-		expect( summaryNumber.props().delta ).toBe( 100 );
-	} );
+		expect( screen.getByText( '1,000.5' ) ).not.toBeEmptyDOMElement();
```
I think replacing `.not.toBeEmptyDOMElement()` with `.toBeInTheDocument()` will make this test more clear. `getBy*` was chosen because the expectation is that the text can be found - the test should fail if it cannot be.

jeffstieler commented on client/marketing/components/recommended-extensions/test/index.js:
```diff
@@ -47,36 +46,49 @@ describe( 'Recommendations and not loading', () => {
-		const spinner = recommendedExtensionsWrapper.find( Spinner );
-		expect( spinner.length ).toBe( 0 );
+		const { container } = recommendedExtensionsWrapper;
+		expect(
+			container.getElementsByClassName( 'components-spinner' )
+		).toHaveLength( 0 );
```
What do you think of:

```js
expect(
	container.querySelector( '.components-spinner' )
).toBeNull();
```

[`querySelector()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector) returns `null` when the selector can't be found.

vbelolapotkov replied:
It will work, but iirc the output is more helpful when `not.toBeInTheDocument()` is used. I'd suggest breaking the test and compare the output. It's fine to go with any option you like more :)

vbelolapotkov commented:
> @vbelolapotkov thank you for the thorough review! (I was only expecting folks on different teams to check the code their team added 😅 )

I wish I understand this before 😜 But hope the feedback was helpful 😄 Good luck!

> I'm going through your feedback now. Thanks again.

👍 you're welcome!

jeffstieler committed:
Use snapshot for testing Leaderboard markup.

jeffstieler committed:
Switch from "not to be empty" to "be in the document".

jeffstieler committed:
No need to waitFor() recordEvent mock.

jeffstieler committed:
Be specific about clicking the "hide" button.

jeffstieler commented on client/wp-admin-scripts/print-shipping-label-banner/setup-notice/test/index.js:
```diff
@@ -10,34 +10,30 @@ import SetupNotice, { setupErrorTypes } from '../';
-		);
-		expect( contents.length ).toBe( 1 );
+	it( 'should show default message for unset error', () => {
+		const { getByText } = render( <SetupNotice isSetupError={ true } /> );
+		expect(
+			getByText(
+				'Unable to set up the plugin. Refresh the page and try again.'
+			)
+		).toHaveClass( 'wc-admin-shipping-banner-install-error' );
```
I defer to @c-shultz here - I just ported team Heisenberg's tests.

jeffstieler committed:
Use toBeVisible() instead of checking style property.

jeffstieler commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/test/index.js:
```diff
@@ -392,47 +480,52 @@ describe( 'The message in the banner', () => {
-		const createShippingLabelText = shippingBannerWrapper.find( 'p' );
-		expect( createShippingLabelText.text() ).toBe( activatedMessage );
+
+		expect(
+			container.querySelector( '.wc-admin-shipping-banner-blob p' )
+				.textContent
+		).toBe( activatedMessage );
```
Also deferring to @c-shultz here - I just ported team Heisenberg's tests.

jeffstieler commented:
Thanks again for the review @vbelolapotkov! I've addressed most of the feedback, and pinged other folks who can weigh in on what's important to test.

If this is looking good though, I think we should get it merged in before there are too many conflicts. 🙃

joelclimbsthings commented:
Hey @jeffstieler , looks great! I'd just been experimenting with `react-testing-library` for the last year, and enthusiastically approve of this change. 

The only thing I'll note is that large snapshots such as `client/analytics/components/leaderboard/test/__snapshots__/index.js.snap` can be a bit of a bad smell at times, and test a lot of implementation details without providing a lot of clarity on the intentions of the developer. Kent Dodds writes a bit about this [here](https://kentcdodds.com/blog/effective-snapshot-testing#things-to-avoid-with-snapshots). Not a big deal in this example, but something to keep in mind.

That said, looks good! :heavy_check_mark:

samueljseay commented:
@joelclimbsthings good point re snapshots. I think we should aim to remove snapshot testing altogether because I don't think we're getting tonnes of value from it. That said, I'd be happy to see that addressed later.

samueljseay reviewed (approved):
Thanks so much for this @jeffstieler 🚢
